src/Storages/Freeze.cpp

```diff
@@ -179,8 +184,13 @@ PartitionCommandsResultInfo Unfreezer::unfreezePartitionsFromTableDirectory(Merg
         for (auto it = disk->iterateDirectory(table_directory); it->isValid(); it->next())
         {
             const auto & partition_directory = it->name();
+            
+            int count_underscores = std::count_if(partition_directory.begin(), partition_directory.end(), []( char c ){return c =='_';});
+            if ((format_version.has_value() && format_version == 0) || count_underscores == 4) {
```

Comment:
Number of underscores can be 4 with new syntax as well (if mutation version is not 0)

Reply:
Also we don't need too parse partition_id at all if `matcher` always returns `true` (when it called from `Unfreezer::unfreeze` or similar place). Currently it can parse some garbage, but simply does not throw exception in this case.

Reply:
I don't understand. Parsing partition_id is needed in order to output it. If I remove check with underscores, 
 - ALTER TABLE UNFREEZE will delete data only in case if format_version is more than zero.
 - SYSTEM UNFREEZE will delete any data despite of format_version
So not checking this leads to different behaviour of these queries. Am I right?

Reply:
Sorry, I missed that SYSTEM UNFREEZE prints partition ids in query result, I though ids are needed only for filtering. So currently it prints invalid ids if table was created with old syntax. It's not clear how to fix it, I see the following options:
 - Do not return list of partitions for SYSTEM UNFREEZE
 - Autodetect format_version somehow for SYSTEM UNFREEZE (not sure if it's possible, but maybe checking some files inside frozen data part can help)

Reply:
Ok, I think not returning list is easier and for sure possible. But still what about deleting data? Different data will be deleted for alter and system queries if not doing autodetect

Reply:
I think it's ok, SYSTEM UNFREEZE will simply remove everything, because it's impossible to distinguish partitions in this case.

Reply:
Hm, it seems like throwing an exception may break some user queries since this logic was introduced for alter unfreeze. May it? Is it critical?https://github.com/ClickHouse/ClickHouse/blob/f60f99d452e9154db8efd8e7fcfc81820f76b2cf/src/Storages/MergeTree/MergeTreeData.cpp#L5250

Reply:
I would rather mark it as TODO and / or create a new issue about this. If I touch this code, the only solution that doesn't break any query I see is to autodetect format version. Otherwise, it may break query and behave differently for alter and system queries. I would create a separate issue for autodetection

Reply:
https://github.com/ClickHouse/ClickHouse/issues/40256